README.md
# DbfEngine Java API

DbfEngine - a Java API to read, write, append and clone xBase(DBASE, Foxpro dbf files). Also API allows read memory files (.mem) of Foxpro.
Samples for both operations you can see at [DbfEngine javadoc](https://smart-flex.ru/htm/de_api/index.html)

This API is pure lightweight library without memory consumption and any third party libraries (there are no java loggers and etc.)
The DBF Java API is intended as a fast engine for data exchange purposes.

## Features

Engine is very small and fast.
API for reading is made as iterator, API for writing is made in manner as JDBC statement. It is allows to write compact Java code.

Also, you can look inside your dbf file by invoking it through command line:
java -jar dbfEngine-1.12.jar your.dbf

The result of parsing dbf header and content will be saved into text file.

## Limitations

This version was tested under MS Foxpro 2.6 without memo field support.

## Requirements

DbfEngine requires JDK 1.6 or higher.

## Code samples
```java
public class Fp26Reader {
    private static void testRead() {
        DbfIterator dbfIterator = DbfEngine.getReader(
            Fp26Reader.class.getResourceAsStream("FP_26_SAMPLE.DBF"), null);

        while (dbfIterator.hasMoreRecords()) {
            DbfRecord dbfRecord = dbfIterator.nextRecord();
            String string = dbfRecord.getString("string");
            float sumFloat = dbfRecord.getFloat("sum_f");
            BigDecimal sumNumeric = dbfRecord.getBigDecimal("sum_n");
            boolean bool = dbfRecord.getBoolean("bool_val");
            Date date = dbfRecord.getDate("date_val");

            System.out.println(string + " " + sumFloat + " " + sumNumeric + " " +
                bool + " " + date);
        }
    }

    public static void main(String[] args) {
        Fp26Reader.testRead();
    }
}

public class Fp26Writer {
    private static void testWrite() {
        DbfAppender dbfAppender = DbfEngine.getWriter("WRT_PERSON.DBF", DbfCodePages.Cp866);

        DbfColumn dc01 = new DbfColumn("magic", DbfColumnTypes.Logical, 0, 0);
        DbfColumn dc02 = new DbfColumn("actor", DbfColumnTypes.Character, 60, 0);
        DbfColumn dc03 = new DbfColumn("currdate", DbfColumnTypes.Date, 0, 0);
        DbfColumn dc04 = new DbfColumn("hit", DbfColumnTypes.Numeric, 10, 2);
        DbfColumn dc05 = new DbfColumn("forever", DbfColumnTypes.Logical, 0, 0);
        dbfAppender.defineColumns(dc01, dc02, dc03, dc04, dc05);

        DbfStatement statement = dbfAppender.getStatement();
        statement.setString("actor", "Chuck Norris");
        statement.setDate("currdate", new Date());
        statement.setBigDecimal("hit", new BigDecimal("500.5"));

        statement.insertStatement();

        statement.setBoolean("magic", Boolean.TRUE);
        statement.setString("actor", "Bruce Lee");
        statement.setBigDecimal("hit", new BigDecimal("1000.10"));
        statement.setBoolean("forever", Boolean.TRUE);

        statement.insertStatement();

        dbfAppender.writeDbfAndClose();
    }

    public static void main(String[] args) {
        Fp26Writer.testWrite();
    }
}

public class DbfClonePerformance {

	private static void cloneDBF(String srcFile, String cloneFile) {
		File room64File = new File(srcFile);
		System.out.println("Length src file (bytes): " + room64File.length() + " (megabytes): " + room64File.length()/(1024*1024));
		DbfHeader header = DbfEngine.getHeader(room64File, "Cp866");
		System.out.println(header.toString());
		DbfIterator dbfIterator = header.getDbfIterator();

		File cloneDbf = new File(cloneFile);
		if (cloneDbf.exists()) {
			cloneDbf.delete();
		}
		DbfAppender dbfAppender = DbfEngine.getWriter(cloneDbf,
				DbfCodePages.Cp866);
		dbfAppender.defineColumns(header);

		while (dbfIterator.hasMoreRecords()) {
			DbfRecord dbfRecord = dbfIterator.nextRecord();

			// here can be your filtering code ....
			
			DbfStatement statement = dbfAppender.getStatement();
			statement.fillStatement(dbfRecord);
			statement.insertStatement();
		}

		dbfAppender.writeDbfAndClose();
		header.closeDbfHeader();
	}

	public static void main(String[] args) {

		long start = System.currentTimeMillis();
		
		cloneDBF("F:\\fias_dbf\\ROOM64.DBF",
				"F:\\fias_dbf\\ROOM64_CLONE.DBF");

		long finish = System.currentTimeMillis();
		
		System.out.println("Clone finished for: " + (finish - start)/1000 + " sec");
	}
}
```
#### Performance

The result of performance are (below is the work log of DbfClonePerformance class):

* Length src file (bytes): 600606501 (megabytes): 572
* DbfHeader [firstRecordPosition=641, countRecords=1078287, countColumns=19, lengthRecord=557, codePage=Cp866, typeDbf={3,FoxBASE_dBASE_III_PLUS_without_memo}]
* Clone finished for: 52 sec

This java code is worked under jdk1.6/32; Intel Core i3; Windows 10. And there is no one messages such as: java.lang.OutOfMemoryError: Java heap space

#### Licensing

DbfEngine is issued on under the GNU Lesser General Public License.

#### Support

If you have any issues or questions or suggestions you can send me a letter by email: <gali.shaimardanov@gmail.com>

#### Related project

Djf is Desktop Java Forms, a compact master-detail UI library like FoxBase, but based on Swing: [You can see Djf here](https://github.com/smart-flex/Djf)



CHANGELOG.md
# Changelog
Version history

|Date      | Ver | Author                      | Description        |
|----------|-----|-----------------------------|--------------------|
|2021-04-25| 1.12| galisha.                    | Added ability for cloning DBF file. |
|2020-08-31| 1.11| galisha.                    | Added ability of append records to existed DBF file. |
|2020-01-08| 1.10| frankvdh, galisha.          | Fixed two issues: 1) Column names not upper case 2) Buffer reads are not stream-safe |
|2019-07-10| 1.09| galisha.                    | Added handling for such bad numeric as: 1219,.0000 |
|2016-10-16| 1.08| galisha.                    | Added validation rules for dbf header in case handling of non-dbf files. |
|2016-07-11| 1.07| galisha.                    | Issued executable command line jar file. It's allows to parse dbf file from OOTB. Fixed error with reading unexpected zero in header of DBF and some code pages. **Sample of usage** : java -jar dbfEngine_bin-1.07.jar your.dbf |
|2015-03-01| 1.06| galisha.                    | Added validation method for dbf header. It's allow to check does this file is DBF file or not. Added method of existing a set of fields in dbf header. |
|2014-01-20| 1.05| galisha.                    | Added into DbfEngine functionality to get dbf header. Added reader for .mem file (new separate package ru.smartflex.tools.dbf.mem) |
|2013-08-26| 1.04| alexei.alexandrov@gmail.com.| Fix bug by reading DBF from zip-archive. (For case when stream is returning less data instead required). |
|2013-07-25| 1.03| galisha.                    | Enhanced exception messages (add current record info). Fix bug with empty value for numerical data. Suppress spaces for empty string. Found Eugene.|
|2013-05-25| 1.02| galisha.                    | Fix bug with reading 1 char in the field. Found Eugene.|
|2013-01-04| 1.01| galisha.                    | Added some code pages|
|2012-12-08| 1.00| galisha.                    | Initial version. Handles MS Foxpro 2.6 dbf file without memo support|


src/main/java/ru/smartflex/tools/dbf/DbfHeader.java
package ru.smartflex.tools.dbf;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/**
 * Contains dbf header definition.
 *
 * @author galisha
 * @since 1.00
 */
public class DbfHeader {

    private int currentOffset = 1;
    private int firstRecordPosition = 0;
    private int countRecords = 0;
    private int typeDbf = -1;
    private int lengthRecord = 0;

    private Map<String, DbfColumn> columns = new HashMap<String, DbfColumn>();
    private String codePage = null;
    private String defaultCodePage = DbfConstants.DEFAULT_CODE_PAGE;
    private InputStream dbfStream = null;
    private List<DbfColumn> orderedColumns = new ArrayList<DbfColumn>();

    private DbfIterator dbfIterator = null;
    private DbfCodePages dbfCodePages = null;

    protected DbfHeader(File dbfFile, String enc, DbfIterator dbfIterator) {
        // Open for reading
        codePageHandling(enc);
        parseDbfHeader(dbfFile);
        this.dbfIterator = dbfIterator;
    }

    protected DbfHeader(InputStream is, String enc, DbfIterator dbfIterator) {
        // Open for reading
        dbfStream = is;
        codePageHandling(enc);
        parseDbfHeader();
        this.dbfIterator = dbfIterator;
    }

    /**
     * Returns iterator.
     *
     * @return iterator.
     * @since 1.05
     */
    public DbfIterator getDbfIterator() {
        return this.dbfIterator;
    }

    /**
     * Closes dbf stream (and also iterator).
     *
     * @see DbfIterator
     * @since 1.05
     */
    public void closeDbfHeader() {
        this.dbfIterator.closeIterator();
    }

    /**
     * Checks column name in dbf header.
     *
     * @param columnName column name
     * @return true if column is existed
     * @since 1.05
     */
    public boolean isColumnExisted(String columnName) {
        boolean fok = false;

        if (columnName != null) {
            String cn = columnName.toUpperCase();
            if (columns.get(cn) != null) {
                fok = true;
            }
        }

        return fok;
    }

    /**
     * Checks entirely array of column names in dbf header.
     *
     * @param columnNames array of column names
     * @return boolean flag
     * @since 1.06
     */
    public boolean isColumnExisted(String... columnNames) {
        boolean fok = false;

        if (columnNames != null && columnNames.length > 0) {
            fok = true;
            for (String name : columnNames) {
                if (columns.get(name.toUpperCase()) == null) {
                    fok = false;
                    break;
                }
            }
        }
        return fok;
    }

    protected DbfHeader() {
        // Create for append
    }

    private void codePageHandling(String enc) {
        if (enc != null) {
            defaultCodePage = enc;
            if (DbfCodePages.getByCharsetName(enc) == null) {
                throw new DbfEngineException(DbfConstants.EXCP_CODE_PAGE + ": " + enc);
            }
        }
    }

    List<DbfColumn> getOrderedColumnList() {
        return orderedColumns;
    }

    /**
     * Returns dbf column iterator.
     *
     * @return dbf column iterator
     * @since 1.05
     */
    public Iterator<DbfColumn> getColumnIterator() {
        return orderedColumns.iterator();
    }

    void addColumn(DbfColumn dbfColumn) {
        if (columns.get(dbfColumn.getColumnName()) != null) {
            throw new DbfEngineException(DbfConstants.EXCP_COLUMN_EXISTED
                    + dbfColumn.getColumnName());
        }
        columns.put(dbfColumn.getColumnName(), dbfColumn);
        orderedColumns.add(dbfColumn);
    }

    DbfColumn getColumn(String name) {
        return columns.get(name);
    }

    private void parseDbfHeader(File dbf) {
        try {
            dbfStream = new FileInputStream(dbf);
            parseDbfHeader();
        } catch (FileNotFoundException e) {
            throw new DbfEngineException(DbfConstants.EXCP_IO_ERROR, e);
        }
    }

    /**
     * Safely reading header from file and network stream; git:frankvdh
     * @return amount of reading bytes
     */
    private int readPartOfHeader(byte[] buff, int offset, int recordLength) throws Exception {
        int ret = recordLength;

        int numLeft = recordLength;
        int offs = offset;
        while (numLeft > 0) {
            int numRead = dbfStream.read(buff, offs, numLeft);
            if (numRead == -1) {
                ret = -1;
                break;
            }
            offs += numRead;
            numLeft -= numRead;
        }
        return ret;
    }

    private void parseDbfHeader() {
        try {
            byte[] hdr = new byte[DbfConstants.DBF_HEADER_LENGTH];
            int cnt = readPartOfHeader(hdr, 0, DbfConstants.DBF_HEADER_LENGTH);
            if (cnt != DbfConstants.DBF_HEADER_LENGTH) {
                dbfStream.close();
                throw new DbfEngineException(DbfConstants.EXCP_HEADER_INF);
            }
            typeDbf = hdr[0] & 0xff;
            // Skip Last update (YYMMDD)
            // Read count of records
            countRecords = hdr[4] & 0xff;
            countRecords += (hdr[5] & 0xff) << 8;
            countRecords += (hdr[6] & 0xff) << 16;
            countRecords += (hdr[7] & 0xff) << 24;
            // Read first record position
            firstRecordPosition = hdr[8] & 0xff;
            firstRecordPosition += (hdr[9] & 0xff) << 8;
            // Read length
            lengthRecord = hdr[10] & 0xff;
            lengthRecord += (hdr[11] & 0xff) << 8;
            // Code page mark
            int cp = hdr[29] & 0xff;
            // System.out.println("cp "+cp);
            if (cp == 0) {
                codePage = defaultCodePage;
                dbfCodePages = DbfCodePages.getByCharsetName(defaultCodePage);
            } else {
                DbfCodePages dcp = DbfCodePages.getByDbfCode(cp);
                if (dcp == null) {
                    throw new DbfEngineException(DbfConstants.EXCP_CODE_PAGE);
                } else {
                    codePage = dcp.getCharsetName();
                    dbfCodePages = dcp;
                }
            }
            int currentOffset = DbfConstants.DBF_HEADER_LENGTH;

            // Fields
            byte[] fld = new byte[DbfConstants.DBF_COLUMN_LENGTH];
            boolean next = true;
            do {
                int flagByte = dbfStream.read();
                if (flagByte == -1) {
                    throw new DbfEngineException(DbfConstants.EXCP_HEADER_END);
                }
                if (flagByte == DbfConstants.DBF_END_HEADER) {
                    next = false;
                }
                currentOffset++;
                if (next) {
                    fld[0] = (byte) flagByte;
                    flagByte = readPartOfHeader(fld, 1, DbfConstants.DBF_COLUMN_LENGTH - 1);
                    if (flagByte == -1) {
                        throw new DbfEngineException(DbfConstants.EXCP_HEADER_END);
                    }
                    DbfColumn dbfColumn = new DbfColumn(this);
                    dbfColumn.parse(fld);
                    // some validation
                    if (dbfColumn.getColumnName() == null) {
                        throw new DbfEngineException(DbfConstants.EXCP_COLUMN_EMPTY);
                    }
                    if (dbfColumn.getDbfColumnType() == null) {
                        throw new DbfEngineException(DbfConstants.EXCP_COLUMN_NOTYPE);
                    }
                    columns.put(dbfColumn.getColumnName(), dbfColumn);
                    orderedColumns.add(dbfColumn);

                    currentOffset += 31;
                }
            } while (next);

            if (currentOffset < firstRecordPosition) {
                do {
                    dbfStream.read();
                    // without any zero analyze yet for e.g. if (flagByte == 0x00)
                    currentOffset++;
                } while (currentOffset < firstRecordPosition);
            }
        } catch (Exception e) {
            throw new DbfEngineException(DbfConstants.EXCP_IO_ERROR, e);
        }
    }

    int getCurrentOffset() {
        return currentOffset;
    }

    void setCurrentOffset(int currentOffset) {
        this.currentOffset = currentOffset;
    }

    /**
     * Returns all amount of records with deleted rows.
     *
     * @return amount of records
     * @since 1.05
     */
    public int getCountRecords() {
        return countRecords;
    }

    String getCodePage() {
        return codePage;
    }

    int getTypeDbf() {
        if (typeDbf == -1) {
            typeDbf = DbfType.FoxBASE_dBASE_III_PLUS_without_memo;
        }
        return typeDbf;
    }

    int getLengthRecord() {
        if (lengthRecord == 0) {
            for (DbfColumn dc : orderedColumns) {
                lengthRecord += dc.getDbfColumnPosition().getColumnLength();
            }
            lengthRecord++; // 1 byte for deleted flag
        }
        return lengthRecord;
    }

    InputStream getDbfStream() {
        return dbfStream;
    }

    /**
     * Returns count of column.
     *
     * @return column count
     * @since 1.05
     */
    public int getCountColumns() {
        return orderedColumns.size();
    }

    /**
     * To String.
     *
     * @since 1.05
     */
    @Override
    public String toString() {
        return "DbfHeader [firstRecordPosition=" + firstRecordPosition
                + ", countRecords=" + countRecords + ", countColumns="
                + getCountColumns() + ", lengthRecord=" + lengthRecord
                + ", codePage=" + codePage + ", typeDbf={" + typeDbf + "," + DbfType.getTypeDescription(typeDbf) + "}]";
    }

    /**
     * Validates dbf header
     *
     * @return true if header is ok
     * @since 1.06
     */
    public boolean isDbfHeaderValid() {
        boolean fok = true;
        if (orderedColumns.size() == 0) {
            fok = false;
        }
        if (fok) {
            for (DbfColumn dc : orderedColumns) {
                if (dc.getColumnName() == null) {
                    fok = false;
                    break;
                }
                if (dc.getDbfColumnType() == null) {
                    fok = false;
                    break;
                }
            }
        }
        return fok;
    }

    /**
     * Gets dbf code page
     * @return dbf code page
     * @since 1.11
     */
    protected DbfCodePages getDbfCodePages() {
        return dbfCodePages;
    }
}
